libsignal-service/src/sender.rs

```diff
@@ -605,53 +617,35 @@ where
         recipient: &ServiceAddress,
         unidentified_access: Option<&SenderCertificate>,
         content: &[u8],
+        is_sync_message: bool,
     ) -> Result<Vec<OutgoingPushMessage>, MessageSenderError> {
         let mut messages = vec![];
-        let mut devices = vec![];
 
-        let myself = *recipient == self.local_address;
-        if !myself || unidentified_access.is_some() {
-            trace!("sending message to default device");
-            messages.push(
-                self.create_encrypted_message(
-                    recipient,
-                    unidentified_access,
-                    DEFAULT_DEVICE_ID.into(),
-                    content,
-                )
-                .await?,
-            );
-        } else {
-            devices.push(DEFAULT_DEVICE_ID);
-        }
+        let mut devices: HashSet<DeviceId> = self
+            .protocol_store
+            .get_sub_device_sessions(recipient)
+            .await?
+            .into_iter()
+            .map(DeviceId::from)
+            .collect();
 
-        devices.extend(
-            self.protocol_store
-                .get_sub_device_sessions(recipient)
-                .await?,
-        );
-        devices.sort_unstable();
-
-        let original_device_count = devices.len();
-        devices.dedup();
-        if devices.len() != original_device_count {
-            log::warn!("SessionStoreExt::get_sub_device_sessions should return unique device id's, and should not include DEFAULT_DEVICE_ID.");
-        }
+        // always send to the primary device no matter what
+        devices.insert(DEFAULT_DEVICE_ID.into());
 
-        // When sending to ourselves, don't include the local device
-        if myself {
-            devices.retain(|id| DeviceId::from(*id) != self.device_id);
+        // when sending a sync message, we always send it to all devices except the current one
+        if is_sync_message {
```

Comment:
Should this be `myself && is_sync_message`? If my memory serves me right, `device_id` is a (small) integer so chances of device numbers colliding is significantly non-zero... Or, `is_sync_message` implies `myself = true` but if so, it would be clearer to have it explicitly stated here.

Reply:
I reworked the PR to not use `is_sync_message` since I finally understood how it's supposed to work.